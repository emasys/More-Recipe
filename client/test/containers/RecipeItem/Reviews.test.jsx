import React from 'react';
import { mount, shallow } from 'enzyme';
import Router from 'react-mock-router';

// Component
import {
  Reviews,
  mapStateToProps
} from '../../../src/containers/RecipeItem/Reviews';
import fakeStore from '../../__mocks__/fakeStore';

let wrapper = null;

jest.mock('../../../src/components/Navbar');
Date.now = jest.fn(() => 1487076708000);

describe('Test suite for Reviews component', () => {
  describe('Render component with one review', () => {
    const props = {
      postReview: jest.fn(),
      recipes: fakeStore.recipes,
      deleteReview: jest.fn(),
      clearReview: jest.fn(),
      review: {
        review: {},
        fetch_reviews: [
          {
            id: 1,
            content: 'new comment',
            createdAt: '2018-03-15T08:03:50.868Z',
            updatedAt: '2018-03-15T08:03:50.868Z',
            recipeId: 1,
            userId: 1,
            reviewsreply: [],
            User: {
              moniker: 'admin',
              avatar: null
            }
          }
        ],
        count_reviews: 1
      },
      auth: fakeStore.user
    };
    beforeEach(() => {
      wrapper = shallow(<Reviews {...props} />);
    });

    it('Should render without errors', () => {
      expect(wrapper.length).toEqual(1);
      expect(wrapper.find('h5.text-center.text-muted').text()).toMatch('Reviews');
      expect(wrapper.find('span.badge.badge-dark').text()).toMatch('1');
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('Should submit a review', () => {
      const instance = wrapper.instance();
      const handleSubmit = jest.spyOn(instance, 'handleForm');
      wrapper
        .find('TextareaAutosize#reviewField')
        .simulate('change', { target: { value: 'A regular food' } });

      wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
      });
      expect(handleSubmit).toBeCalled();
    });

    it('should delete a review and unmount', () => {
      const app = mount(<Router>
        <Reviews {...props} />
      </Router>);
      app.find('i.delete-review-btn').simulate('click');
      app.unmount();
      expect(props.clearReview).toBeCalled();
    });
    it('should call componentWillReceiveProps', () => {
      const instance = wrapper.instance();
      const compWRP = jest.spyOn(instance, 'componentWillReceiveProps');
      const loadMore = jest.spyOn(instance, 'loadMoreReviews');
      const newProps = {
        postReview: jest.fn(),
        recipes: fakeStore.recipes,
        deleteReview: jest.fn(),
        getReviews: jest.fn(),
        review: {
          review: {},
          fetch_reviews: [
            {
              id: 1,
              content: 'new comment',
              createdAt: '2018-03-15T08:03:50.868Z',
              updatedAt: '2018-03-15T08:03:50.868Z',
              recipeId: 1,
              userId: 1,
              User: {
                moniker: 'admin',
                avatar: null
              }
            },
            {
              id: 2,
              content: 'new comment',
              createdAt: '2018-03-15T08:03:50.868Z',
              updatedAt: '2018-03-15T08:03:50.868Z',
              recipeId: 1,
              userId: 1,
              User: {
                moniker: 'admin',
                avatar: null
              }
            }
          ],
          count_reviews: 2
        },
        auth: fakeStore.user
      };
      wrapper.setProps(newProps);
      expect(compWRP).toHaveBeenCalled();
      expect(instance.state.showMore).toBe(true);
      wrapper
        .find('button.btn.mt-10.btn-lg.btn-block.bg-light')
        .simulate('click', {
          preventDefault: () => {}
        });
      expect(loadMore).toHaveBeenCalled();

      wrapper.setState({ offset: 3 });
      wrapper.setProps(newProps);
      expect(compWRP).toHaveBeenCalled();
      expect(instance.state.showMore).toBe(false);
    });

    it('should test mapStateToProps', () => {
      const initialState = {
        review: {},
        recipes: {},
        user: {}
      };
      expect(mapStateToProps(initialState).auth).toEqual({});
      expect(mapStateToProps(initialState).recipes).toEqual({});
      expect(mapStateToProps(initialState).review).toEqual({});
    });
  });
});
