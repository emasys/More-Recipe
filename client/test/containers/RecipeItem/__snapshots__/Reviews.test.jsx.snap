// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for Reviews component Render component with one review should match snapshot 1`] = `
<div
  className="row justify-content-center mt-2 catalog-wrapper"
>
  <div
    className="col-12"
  >
    <h5
      className="text-center text-muted"
    >
      Reviews
       
      <span
        className="badge badge-dark"
        data-tip="Total number of reviews"
      >
        1
      </span>
    </h5>
    <hr />
  </div>
  <div
    className="col-lg-7 col-sm-12 recipe-image"
  >
    <form
      className="text-center"
      onSubmit={[Function]}
    >
      <div
        className="form-row"
      >
        <div
          className="form-group col-md-12 col-sm-12"
        >
          <TextareaAutosize
            className="textarea col-12"
            id="reviewField"
            name="reviewBox"
            onChange={[Function]}
            onHeightChange={[Function]}
            placeholder="leave a review"
            useCacheForDOMMeasurements={false}
            value=""
          />
        </div>
      </div>
      <button
        className="btn btn-dark  btn-lg col-lg-4 col-md-6 col-sm-12"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
  <div
    className="col-lg-8 col-sm-12 "
  >
    <GenerateReviews
      auth={
        Object {
          "allUsers": Object {
            "success": true,
            "users": Array [
              Object {
                "avatar": null,
                "bio": null,
                "country": null,
                "createdAt": "2018-03-13T05:04:37.443Z",
                "email": "emasysnd@gmail.com",
                "firstName": null,
                "id": 1,
                "lastName": null,
                "moniker": "admin",
                "password": "$2a$10$47oncNYPerLpBX2BfrUSpeUo4KMIU5dsHkJo3hRUWw1n58K6muoUS",
                "updatedAt": "2018-03-13T05:04:37.443Z",
              },
            ],
          },
          "authInfo": Object {
            "isLoggedIn": true,
            "userId": 1,
            "username": "admin",
          },
          "isLoggedIn": true,
          "message": null,
          "userInfo": Object {
            "data": Object {
              "avatar": null,
              "bio": null,
              "country": "Nigeria",
              "email": "emasysnd@gmail.com",
              "firstName": "emmy",
              "id": 1,
              "lastName": "endy",
              "moniker": "admin",
            },
            "success": true,
          },
          "userProfile": Object {
            "data": Object {
              "avatar": null,
              "bio": null,
              "country": null,
              "email": "emasysnd@gmail.com",
              "firstName": null,
              "id": 1,
              "lastName": null,
              "moniker": "admin",
            },
            "success": true,
          },
        }
      }
      deleteReview={[Function]}
      review={
        Object {
          "count_reviews": 1,
          "fetch_reviews": Array [
            Object {
              "User": Object {
                "avatar": null,
                "moniker": "admin",
              },
              "content": "new comment",
              "createdAt": "2018-03-15T08:03:50.868Z",
              "id": 1,
              "recipeId": 1,
              "updatedAt": "2018-03-15T08:03:50.868Z",
              "userId": 1,
            },
          ],
          "review": Object {},
        }
      }
    />
  </div>
</div>
`;
